### PR TITLE
feature/news

### PR DESCRIPTION
**Pull request by haritonasty**
**Created News component for TrendsExplorePage**

image

image

### CODE REVIEW

File: src/components/News/News.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import React from 'react'
+import NewsCard from './NewsCard'
+import styles from './News.module.scss'
+
+const News = ({ data }) => {
+  return (
+    <div className={styles.items}>
+      {data.map(item => (
+        <NewsCard
+          key={`${item.title} ${item.description}`}
```
Comment (Partysun) [inline src/components/News/News.js]: Why so hard logic for a key. If items will not be changed, you can use index without any performance issues. @DmitriVanGuard Your mention?

File: src/components/News/News.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import React from 'react'
+import NewsCard from './NewsCard'
+import styles from './News.module.scss'
+
+const News = ({ data }) => {
```
Comment (Partysun) [inline src/components/News/News.js]: What is the default value for data?

File: src/components/Trends/Explore/TrendsExploreAdditionalInfo.js [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+  }
+
+  const tabs = []
+  if (insights.length > 0) tabs.push(insightsTab)
+  if (news.length > 0) tabs.push(newsTab)
+
+  let defaultSelectedTab
+
+  if (tabs.length === 0) return null
+  else defaultSelectedTab = tabs[0].index
```
Comment (Partysun) [inline src/components/Trends/Explore/TrendsExploreAdditionalInfo.js]: @haritonasty Do we really need "else" here?
If tabs.length === 0 ok. return null.
After const defaultSelectedTab = tabs[0].index

File: src/components/Trends/Explore/TrendsExploreAdditionalInfo.js [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+              />
+            ))}
+          </div>
+        )}
+      </div>
+    </section>
+  )
+}
+
+const getTrendsTags = numberOfLastDays => {
```
Comment (Partysun) [inline src/components/Trends/Explore/TrendsExploreAdditionalInfo.js]: @haritonasty You can write it without an imperative style of code.
return Array.from({length: numberOfLAstDays}, (_, index) => additionalLogic(index))

File: src/components/News/News.module.scss [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  @include responsive('desktop', 'laptop') {
+    &:nth-child(3n) {
+      margin-right: 0;
+    }
+  }
+
+  @include responsive('tablet') {
+    width: 49%;
+    margin-bottom: 2%;
+    margin-right: 2%;
```
Comment (DmitriVanGuard) [inline src/components/News/News.module.scss]: Already declared on the LOC `18,19`

File: src/components/MultilineText/MultilineText.js [hunk 6]
```diff
@@ -59,7 +59,7 @@ class MultilineText extends React.PureComponent {
     const oneLineHeight = getOneLineHeight(id)
     this.oneLineHeight = oneLineHeight
 
-    if (Math.floor(container.offsetHeight / oneLineHeight) > maxLines) {
+    if (Math.ceil(container.offsetHeight / oneLineHeight) > maxLines) {
```
Comment (DmitriVanGuard) [inline src/components/MultilineText/MultilineText.js]: Why `ceil`? What's the edge case?

File: src/components/Insight/InsightsTrends.js [hunk 7]
```diff
@@ -17,6 +17,8 @@ const InsightsTrends = ({ allInsightsByTag, ...props }) => {
   )
 }
 
+export const oneDayTimestamp = 1000 * 60 * 60 * 24
```
Comment (DmitriVanGuard) [inline src/components/Insight/InsightsTrends.js]: Use [this const](https://github.com/santiment/app/blob/9c0007579f7004e828ad668ef8dc78c24b9ba3d8/src/utils/dates.js#L4) from a `utils/dates`

File: src/components/Trends/Explore/TrendsExploreAdditionalInfo.js (hunk 4 above)
Comment (DmitriVanGuard) [reply src/components/Trends/Explore/TrendsExploreAdditionalInfo.js]: @haritonasty We have this function somewhere, just import it

File: src/components/Trends/Explore/TrendsExploreAdditionalInfo.js [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+const enhance = compose(
+  ...getPast3DaysInsightsByTrendTag(),
+  graphql(NEWS_QUERY, {
+    options: ({ word: tag }) => {
+      const { from, to } = getTimeIntervalFromToday(-3, DAY)
+      return {
+        variables: { from, to, tag, size: 6 }
+      }
+    },
+    props: ({ data: { news = [] } }) => ({ news: news.reverse() })
```
Comment (DmitriVanGuard) [inline src/components/Trends/Explore/TrendsExploreAdditionalInfo.js]: @haritonasty Is it guaranteed that news are returned in the chronological (ascending) order?

File: src/components/MultilineText/MultilineText.js (hunk 6 above)
Comment (haritonasty) [reply src/components/MultilineText/MultilineText.js]: `Math.floor(79/40) = 1`
`Math.ceil(79/40) = 2`

were problems with this behavior in news @DmitriVanGuard

File: src/components/Trends/Explore/TrendsExploreAdditionalInfo.js (hunk 4 above)
Comment (haritonasty) [reply src/components/Trends/Explore/TrendsExploreAdditionalInfo.js]: @DmitriVanGuard didn't find mentions `Array.from`... Create it in `utils.js`?

File: src/components/Trends/Explore/TrendsExploreAdditionalInfo.js (hunk 8 above)
Comment (haritonasty) [reply src/components/Trends/Explore/TrendsExploreAdditionalInfo.js]: @DmitriVanGuard "graphql wrapper is sorting them by timestamp" - so, yes

File: src/components/Trends/Explore/TrendsExploreAdditionalInfo.js (hunk 4 above)
Comment (DmitriVanGuard) [reply src/components/Trends/Explore/TrendsExploreAdditionalInfo.js]: Export and use this function

https://github.com/santiment/app/blob/93632350aab667c249dfcc3b290f88a3f2d570f4/src/components/Insight/InsightsTrends.js#L22

File: src/components/Trends/Explore/TrendsExploreAdditionalInfo.js [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+  )
+}
+
+const getTrendsTags = numberOfLastDays =>
+  Array.from({ length: numberOfLastDays }, (_, idx) =>
+    getInsightTrendTagByDate(new Date(Date.now() - ONE_DAY_IN_MS * idx))
+  )
+
+const filterInsights = (insights = [], word) =>
+  insights.filter(({ tags }) => tags.find(({ name }) => name === word))
```
Comment (DmitriVanGuard) [inline src/components/Trends/Explore/TrendsExploreAdditionalInfo.js]: `.find` returns found item if it exists. In our case, we just need to get a boolean value -> `.some`

File: src/components/Trends/Explore/TrendsExploreAdditionalInfo.js (hunk 8 above)
Comment (DmitriVanGuard) [reply src/components/Trends/Explore/TrendsExploreAdditionalInfo.js]: Can't find where it is stated. Give me a link, please 🙏

Comment (haritonasty) [reply src/components/Trends/Explore/TrendsExploreAdditionalInfo.js]: oh, it in my private dialog with Tsetso ))

image

**Pull request merged**